### PR TITLE
Datepicker using Globalize 1.x

### PR DESCRIPTION
**Pull request by rxaviers**
:boom: WIP...

The goal is to have Datepicker working with Globalize 1.x.

Depends on
- [x] https://github.com/jquery/jquery-ui/pull/1371 `fnagel:calendar-value-pr3`;

Todo
- [ ] Rebase on datepicker branch
- [x] Destroy translation data (weekHeader, prevText, nextText , datePickerRole, currentText). Create user options instead;
- [x] Move `external/date.js` into `ui/` (see #1260 (comment));
- [ ] Fix RTL LTR direction (available at `cldr.main("layout/orientation/characterOrder")`, values are `left-to-right` or `right-to-left`);
- [ ] Get currentText: "Today" from locale.
- [x] Fix Use narrow day names (e.g., "ﺥ" EEEEE) if short (e.g., "ﺎﻠﺨﻤﻴﺳ" EEE) is too lengthy;
- [ ] Update all demos accordingly. (currently, demos/datepicker/localization.html has been updated)

### CODE REVIEW

**Comment by fnagel** [comment]
@rxaviers #1371 has been merged!

**Comment by rxaviers** [comment]
:tada: :tada:

File: ui/calendar/date.js [hunk 1]
```diff
@@ -17,40 +36,47 @@ var weekdays = [ "sun", "mon", "tue", "wed", "thu", "fri", "sat" ],
...
 		this.dateObject = new Date( time );
 		return this;
 	},
+
 	setDay: function( day ) {
 		var date = this.dateObject;
+		// FIXME: Why not to use .setDate?
 		this.dateObject = new Date( date.getFullYear(), date.getMonth(), day, date.getHours(),
 			date.getMinutes(), date.getSeconds() );
 		return this;
```
Comment (jzaefferer) [inline ui/calendar/date.js]: This drops the milliseconds, but I doubt that's intentional. Let's fix it.

**Comment by jzaefferer** [comment]
This still has many commits from the other PR, does this need a rebase to get rid of them?

**Comment by jzaefferer** [comment]
As for RTL: Scott is putting together a meeting to discuss a UI-wide solution. @rxaviers should be there as well and provide some input from the CLDR perspective (what RTL data is provided by CLDR?). Since that might still take a while, I suggest ignoring RTL here and addressing that later.

File: ui/calendar/date.js (hunk 1 above)
Comment (fnagel) [inline ui/calendar/date.js]: Not sure if this was needed on order to prevent referencing of the new variable.

**Comment by fnagel** [comment]
@jzaefferer Yes, this will need another rebase as we chaged a bunch of older commits in the merged branch.

File: ui/datepicker.js [hunk 2]
```diff
@@ -14,35 +14,34 @@
...
 
-// TODO use uniqueId, if possible
-var idIncrement = 0,
-	// TODO move this to the instance
-	suppressExpandOnFocus = false;
+var widget,
 
-$.widget( "ui.datepicker", {
+	// FIXME can we make this dynamic?
+	calendarOptions = [ "buttons", "disabled", "locale", "eachDay", "max", "min", "numberOfMonths", "showWeek" ];
```
Comment (fnagel) [inline ui/datepicker.js]: See PR #1432

**Comment by fnagel** [comment]
@rxaviers Some questions regarding this branch / the implementation of Globalize:

* I noticed that the first two days are the same (Sunday or Monday depending on locale). Can you give me a hint where to search for the issue?
* What needs to be done to update to Globalize 1.0.0 (its alpha 7 now)? I've tried (using bower, adding globalize number) but run into a cldr data error.

**Comment by fnagel** [comment]
My current working branch: https://github.com/fnagel/jquery-ui/tree/datepicker-globalize

Not sure if I should force push this PR or open a new one with reference and copied description. Opinions? @rxaviers @scottgonzalez @jzaefferer

**Comment by scottgonzalez** [comment]
A force push seems fine for this one.

**Comment by rxaviers** [comment]
> I noticed that the first two days are the same (Sunday or Monday depending on locale). Can you give me a hint where to search for the issue?

What do you mean?

**Comment by rxaviers** [comment]
> What needs to be done to update to Globalize 1.0.0 (its alpha 7 now)? I've tried (using bower, adding globalize number) but run into a cldr data error. 

I need a little more details.

**Comment by rxaviers** [comment]
About passing i18n options to datepicker (e.g., `locale`, `pattern`, etc), I suggest the below (we had a discussion about it in the past, but I don't find a summary about it; anyway...)

Datepicker shouldn't be tight coupled with Globalize. Instead, it should expect formatters and parsers (functions needed for formatting and parsering the calendar and input values) to be passed from user land. Datepicker could use Globalize to fill in the default ones (i.e., for `en`). Datepicker could even have a helper that uses Globalize to automatically fill them in for other locales as well. An initial work started at a8d5e80c12e2ddbd315c2c1374e30319a294f88c (note `_calendarDateOptions`). But, I would consider removing `_setLocale` and any `var globalize = new Globalize( locale )` reference from within datepicker and expect all formatters and parsers to come from `options`.

Having Datepicker loosely coupled with Globalize the way I described may sound more work now. But, it will dramatically improve its maintenance over time + it will help people to integrate datepicker into their i18n workflow easily.

**Comment by rxaviers** [comment]
> Not sure if I should force push this PR

Feel free to force push to it assuming no work here is lost. :)

**Comment by scottgonzalez** [comment]
I don't recall coming to that conclusion. That's the approach we agreed on for spinner, but I have always expected datepicker to have a hard dependency on Globalize.

**Comment by fnagel** [comment]
> > I noticed that the first two days are the same (Sunday or Monday depending on locale). Can you give me a hint where to search for the issue?

> What do you mean?

first-day-twice

**Comment by rxaviers** [comment]
Well remembered @scottgonzalez. Obviously, having datepicker to have a hard dependency on Globalize is up to you. I also remember that you wanted datepicker to be distributed working for `en` as a default, which needs to be taken into account in this implementation.

**Comment by rxaviers** [comment]
@fnagel related code:
- https://github.com/jquery/jquery-ui/pull/1341/files#diff-b2fccddb29452175f8032dd5ae8ccb61R170
- https://github.com/jquery/jquery-ui/pull/1341/files#diff-563a5392a566d1a29a68b1b5a37c4cb6R168
- https://github.com/jquery/jquery-ui/pull/1341/files#diff-563a5392a566d1a29a68b1b5a37c4cb6R162

Did I answer your question? Please, just let me know on any problem. Also, follow a code that may help you to play with Globalize: https://gist.github.com/rxaviers/1da4a40b13d2a6b03316

**Comment by dwachss** [comment]
Thank you for rewriting Datepicker with the widget factory and based on CLDR. 
Two questions about where this is going:
1. Will there be support for calendar systems other than gregorian? CLDR includes the data for localization but not the algorithms. Will Datepicker be extensible that way?
2. The initial comment 'Get currentText: "Today" from locale'. That's not in CLDR as far as I can tell (can you get that from Relative Date Names?). What about the "Done" button--or will you leave that out?

**Comment by fnagel** [comment]
> 1. Will there be support for calendar systems other than gregorian? CLDR includes the data for localization but not the algorithms. Will Datepicker be extensible that way?

Afaik yes.

> What about the "Done" button--or will you leave that out?

There is a new option buttons like in dialog -- you will need to set the text strings yourself. See https://github.com/jquery/jquery-ui/blob/datepicker/demos/calendar/buttonbar.html

Make sure to take a look at the wiki page: http://wiki.jqueryui.com/w/page/12137778/Datepicker

**Comment by rxaviers** [comment]
> 1. Will there be support for calendar systems other than gregorian? CLDR includes the data for localization but not the algorithms. Will Datepicker be extensible that way?

I can't answer for the widget. But, it will depend on Globalize, whose support for other calendars need https://github.com/jquery/globalize/issues/320 to be implemented. Help appreciated.

> 2. The initial comment 'Get currentText: "Today" from locale'. That's not in CLDR as far as I can tell (can you get that from Relative Date Names?)

It's is in CLDR (in [dateFields.json](http://www.unicode.org/repos/cldr-aux/json/26/main/en/dateFields.json)).

**Comment by dwachss** [comment]
Thanks. I've been working on an alternative-calendar-system datepicker widget for the past few years ( https://github.com/dwachss/flexcal ) and I have a good handle on manipulating them. I'll see if I have time to work on Globalize.
Thanks also for the link to the CLDR data (it's very hard to find from the Unicode site).
Danny   


     On Thursday, April 30, 2015 10:53 AM, Rafael Xavier de Souza <notifications@github.com> wrote:
   

 
   
   - Will there be support for calendar systems other than gregorian? CLDR includes the data for localization but not the algorithms. Will Datepicker be extensible that way?

I can't answer for the widget. But, it will depend on Globalize, whose support for other calendars need jquery/globalize#320 to be implemented. Help appreciated.
   
   - The initial comment 'Get currentText: "Today" from locale'. That's not in CLDR as far as I can tell (can you get that from Relative Date Names?)

It's is in CLDR (in dateFields.json).—

**Comment by rxaviers** [comment]
Thanks @dwachss. Looking forward to it.

**Comment by fnagel** [comment]
Current dev branch (merged with @scottgonzalez latest changes to datepicker branch):
https://github.com/fnagel/jquery-ui/commits/datepicker-globalize

**Comment by rxaviers** [comment]
Can this branch `datepicker-globalize-1.x` be deleted?

In the description, there are two remaining items. Are they been tracked somewhere else?

**Comment by fnagel** [comment]
I closed this by accident. It's not merged to datepicker branch yet. 

datepicker-globalize-1.x is still up to date with my latest changes!

**Comment by rxaviers** [comment]
Gotcha. Thanks

**Comment by fnagel** [comment]
More info in the UI dev meeting later today.

**Comment by fnagel** [comment]
This PR has been merged to datepicker branch. Remaining issue is RTL LTR support.

@scottgonzalez Datepicker branch is now up to date and waiting for your merging magic :-D

**Comment by scottgonzalez** [comment]
I finished merging master into datepicker this morning.

**Comment by fnagel** [comment]
Looks good! Added a new PR based on this: #1590

**Pull request merged**